lib/sycamore/sycamore/docset.py

```diff
@@ -1407,21 +1407,10 @@ def top_k(
     def llm_generate_group(self, llm: LLM, instruction: str, field: str, **kwargs):
         # Not all documents will have a value for the given field, so we filter those out.
 
-        def filter_meta(row):
-            doc = Document.from_row(row)
-            return not isinstance(doc, MetadataDocument)
-
-        def extract_entities(row):
-            doc = Document.from_row(row)
-            value = doc.field_to_value(field)
-            return {"key": value}
-
-        values = self.plan.execute().filter(filter_meta).map(extract_entities).materialize()
-
-        count = values.count()
-        samples = values if count < 1000 else values.random_sample(1000.0 / count)
+        count = self.count()
```

Comment:
I think this code is going to force execution twice. 
You don't know the count unless you execute the entire pipeline. I think that self.count() will run through the standard Executor so will materialize (if that's enabled) and at least not re-do the full effort.  

A more efficient implementation would be to do a streaming read and use the choose k out of n algorithm from https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle; so you store the 1000 documents, and process all of them exactly once.   I'd probably just put this in as a TODO.

Reply:
Remember to add the TODO.